src/source/source_cache.js

```diff
@@ -52,6 +54,7 @@ class SourceCache extends Evented {
     transform: Transform;
     _isIdRenderable: (id: number) => boolean;
     used: boolean;
+    _state: SourceFeatureState
```

Comment:
I think it does make sense to _store_ a `SourceFeatureState` here on the `SourceCache`, since it belongs to a source and should be discarded precisely when a source is discarded.

But I agree that the state-aware _logic_ in below feels a bit out of place here.  What if we:
 - move the responsibility for calling `Tile#updateFeatureState` during `prepare()` directly into `SourceFeatureState#coalesceChanges()` (i.e., take a set of tiles as a parameter to `coalesceChanges`)?
 - Add a `SourceFeatureState#initializeTileState(Tile)` method and replace the other two cases of `tile.updateFeatureState(this._state.state);` with `this._state.initializeTileState(tile)`?

Reply:
> are there any alternatives to storing state for all features on all loaded tiles in the `SourceCache`?

Since `SourceCache` is responsible for creating sources and managing their tiles, it is the appropriate place to manage the feature state of the source as well. 
@anandthakker, that same reasoning is what led me to make `SourceFeatureState` a data-only class and have `SourceCache`  continue to be responsible for managing tiles and their state. 

How about adding `SourceCache#initializeTile`, called for newly loaded tiles, or tiles fetched from the cache? It would update feature state and backfill raster DEM tiles.

Reply:
> Since SourceCache is responsible for creating sources and managing their tiles, it is the appropriate place to manage the feature state of the source as well

I disagree.  `SourceCache` is already very difficult to reason about, in part because it's doing too much. (See also https://github.com/mapbox/mapbox-gl-js/issues/2291.)  Ideally, it would not manipulate the "content" of its `Tile`s at all (except perhaps setting `tile.state`, since that's directly related to its job of caching tiles).